### ISSUE TITLE
make  error .

### ISSUE THREAD

**Issue by w796933** [simon987/ngx_http_js_challenge_module#2]
sir   make   error . 
	-o objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o \
	/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c: In function ‘get_challenge_string’:
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:261:5: error: implicit declaration of function ‘SHA1’ [-Werror=implicit-function-declaration]
     SHA1((unsigned char *) buf, (size_t) (sizeof(int32_t) + addr.len + secret.len), md);
     ^~~~
cc1: all warnings being treated as errors
objs/Makefile:1203: recipe for target 'objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o' failed
make[1]: *** [objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o] Error 1
make[1]: Leaving directory '/root/nginx-1.16.1'
Makefile:8: recipe for target 'build' failed
make: *** [build] Error 2

**Comment by simon987** [comment]
Hi, can you give me the output of `nginx -V` please?

**Comment by w796933** [comment]
before  error ， i fix 。 but   not this error

**Comment by w796933** [comment]
/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c: In function ‘get_cookie’:
/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c:306:5: error: ‘for’ loop initial declarations are only allowed in C99 mode
     for (ngx_uint_t i = 0; i < r->headers_in.cookies.nelts; i++) {
  
how to  use  c99  mode  make ?

**Comment by w796933** [comment]
this  error  is fix  
 just add

CFLAGS=-std=c99

To the top of your makefile, or you can make the code compliant with C90 (as LukeN suggests.)

**Comment by simon987** [comment]
I modified the build script with the `CFLAGS=-std=c99` variable, thank you

**Comment by w796933** [comment]
use  c99 mode   i  have  make  no pass 


[root@nginx-test nginx-1.16.1]# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-39) (GCC) 




[root@nginx-test nginx-1.16.1]# ./configure --user=www --group=www --prefix=/usr/local/nginx --with-http_stub_status_module --with-http_ssl_module --with-http_v2_module --with-http_gzip_static_module --with-http_sub_module --with-openssl=openssl-1.1.1d --add-module=/root/ngx_http_js_challenge_module




/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c: In function ‘ngx_http_js_challenge_handler’:
src/core/ngx_string.h:40:57: error: expected expression before ‘)’ token
 #define ngx_string(str)     { sizeof(str) - 1, (u_char *) str }
                                                         ^
/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c:344:43: note: in expansion of macro ‘ngx_string’
     int ret = get_cookie(r, &((ngx_str_t) ngx_string("res")), &response);
                                           ^
src/core/ngx_string.h:40:56: error: invalid operands to binary * (have ‘struct ngx_command_t *’ and ‘struct ngx_command_t *’)
 #define ngx_string(str)     { sizeof(str) - 1, (u_char *) str }
                                                        ^
/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c:344:43: note: in expansion of macro ‘ngx_string’
     int ret = get_cookie(r, &((ngx_str_t) ngx_string("res")), &response);
                                           ^
src/core/ngx_string.h:40:57: warning: initialization from incompatible pointer type [enabled by default]
 #define ngx_string(str)     { sizeof(str) - 1, (u_char *) str }
                                                         ^
/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c:344:43: note: in expansion of macro ‘ngx_string’
     int ret = get_cookie(r, &((ngx_str_t) ngx_string("res")), &response);
                                           ^
src/core/ngx_string.h:40:57: warning: (near initialization for ‘(anonymous).data’) [enabled by default]
 #define ngx_string(str)     { sizeof(str) - 1, (u_char *) str }
                                                         ^
/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c:344:43: note: in expansion of macro ‘ngx_string’
     int ret = get_cookie(r, &((ngx_str_t) ngx_string("res")), &response);
                                           ^
src/core/ngx_string.h:40:57: warning: passing argument 2 of ‘get_cookie’ from incompatible pointer type [enabled by default]
 #define ngx_string(str)     { sizeof(str) - 1, (u_char *) str }
                                                         ^
/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c:344:43: note: in expansion of macro ‘ngx_string’
     int ret = get_cookie(r, &((ngx_str_t) ngx_string("res")), &response);
                                           ^
/root/ngx_http_js_challenge_module/ngx_http_js_challenge.c:301:5: note: expected ‘struct ngx_str_t *’ but argument is of type ‘struct ngx_command_t (*)[1]’
 int get_cookie(ngx_http_request_t *r, ngx_str_t *name, ngx_str_t *value) {
     ^
make[1]: *** [objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o] Error 1
make[1]: Leaving directory `/usr/local/src/nginx-1.16.1'
make: *** [build] Error 2

**Comment by simon987** [comment]
@w796933 please let me know if you can build with the latest commit

**Comment by w796933** [comment]
yes . latest  commit  build

**Comment by w796933** [comment]
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:54:17: error: initializer element is not constant
                 str_js_challenge,
                 ^
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:54:17: error: (near initialization for ‘ngx_http_js_challenge_commands[0].name’)
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:62:17: error: initializer element is not constant
                 str_js_challenge_bucket_duration,
                 ^
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:62:17: error: (near initialization for ‘ngx_http_js_challenge_commands[1].name’)
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:70:17: error: initializer element is not constant
                 str_js_challenge_bucket_duration,
                 ^
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:70:17: error: (near initialization for ‘ngx_http_js_challenge_commands[2].name’)
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:78:17: error: initializer element is not constant
                 str_js_challenge_html,
                 ^
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:78:17: error: (near initialization for ‘ngx_http_js_challenge_commands[3].name’)
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:86:17: error: initializer element is not constant
                 str_js_challenge_title,
                 ^
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:86:17: error: (near initialization for ‘ngx_http_js_challenge_commands[4].name’)
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c: In function ‘get_cookie’:
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:303:5: error: ‘for’ loop initial declarations are only allowed in C99 mode
     for (ngx_uint_t i = 0; i < r->headers_in.cookies.nelts; i++) {
     ^
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:303:5: note: use option -std=c99 or -std=gnu99 to compile your code
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c: At top level:
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:47:18: error: ‘str_js_challenge_secret’ defined but not used [-Werror=unused-variable]
 static ngx_str_t str_js_challenge_secret = ngx_string("js_challenge_secret");
                  ^
cc1: all warnings being treated as errors
make[1]: *** [objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o] Error 1
make[1]: Leaving directory `/usr/local/src/nginx-1.16.1'
make: *** [modules] Error 2
[root@nginx-test ngx_http_js_challenge_module]# cat build.sh 
#!/bin/bash

if [ -z ${NGINX_PATH+x} ]; then
  echo "Please set the NGINX_PATH variable";
  exit
fi

MODULE_PATH=$(pwd)/
CONFIG_ARGS=$(nginx -V 2>&1 | tail -n 1 | cut -c 21- | sed 's/--add-dynamic-module=.*//g')
CONFIG_ARGS="${CONFIG_ARGS} --add-dynamic-module=${MODULE_PATH}"
echo $CONFIG_ARGS

(
  cd ${NGINX_PATH} || exit
  bash -c "./configure ${CONFIG_ARGS}"
  CFLAGS=-std=c99 make modules -j "$(nproc)"
)

**Comment by simon987** [comment]
Sorry about that, can you try again with the latest commit? I can't really test it myself because I don't have the same GCC version

**Comment by w796933** [comment]
my build os is  centos 7  64bit  

you build os is ? 

i try other  os  build it  .

**Comment by simon987** [comment]
I am building it with Debian testing (`Debian 5.4.19-1 (2020-02-13) x86_64 GNU/Linux`)

**Comment by w796933** [comment]
Debian 5.4.19-1  is  to old   os .

**Comment by simon987** [comment]
`5.4.19-1` is the kernel version, I am using *"bullseye"* (Debian 11)

**Comment by w796933** [comment]
okay . i   test  debian 11 .   so thanks .

**Comment by simon987** [comment]
I made the code C90 compatible, with the last commit, you should no longer see the C99 errors. Please let me know if this works

**Comment by w796933** [comment]
okay . i test it . 
how can auto   drop    JavaScript challenge   no pass ip  address .

**Comment by simon987** [comment]
> how can auto drop JavaScript challenge no pass ip address .

Can you reformulate the question? I don't understand what you mean at all

**Comment by w796933** [comment]
js verification failed attack IP. How to automatically block access   ？  how  to use   iptables  auto  drop  access ?

**Comment by simon987** [comment]
>  how to use iptables auto drop access ?
I'm not sure, you would need to ask someone who is more experienced than me with iptables.

**Comment by w796933** [comment]
i use   debian 11  run  build.sh   this error . 


make -f objs/Makefile modules
make[1]: Entering directory '/root/nginx-1.16.1'
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
	-o objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o \
	/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c: In function ‘get_challenge_string’:
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:253:5: error: implicit declaration of function ‘SHA1’ [-Werror=implicit-function-declaration]
     SHA1((unsigned char *) buf, (size_t) (sizeof(int32_t) + addr.len + secret.len), md);
     ^~~~
cc1: all warnings being treated as errors
make[1]: *** [objs/Makefile:1219: objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o] Error 1
make[1]: Leaving directory '/root/nginx-1.16.1'
make: *** [Makefile:14: modules] Error 2

**Comment by simon987** [comment]
@w796933 please let me know if you still have this error after the last commit

**Comment by w796933** [comment]
use  last commit   the some  error  .



make -f objs/Makefile modules
make[1]: Entering directory '/root/nginx-1.16.1'
cc -c -fPIC -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
	-o objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o \
	/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c: In function ‘verify_response’:
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:287:5: error: implicit declaration of function ‘SHA1’ [-Werror=implicit-function-declaration]
     SHA1((unsigned char *) response.data, response.len, md);
     ^~~~
cc1: all warnings being treated as errors
make[1]: *** [objs/Makefile:1219: objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o] Error 1
make[1]: Leaving directory '/root/nginx-1.16.1'
make: *** [Makefile:14: modules] Error 2

**Comment by simon987** [comment]
Sorry about that.
Can you retry with the latest commit?

**Comment by w796933** [comment]
sir . how to  add  white  ip   or url   or  ua  ?

**Comment by simon987** [comment]
> how to add white ip or url or ua ?

You can do that with basic nginx configuration, this is out of the scope of this project

**Comment by w796933** [comment]
i  mean  how to add white ip or url or ua   . because  if   enabled  js  challenge module ,  will deny   app   load .  and  deny   search engines to crawl.

**Comment by w796933** [comment]
The compilation has been unsuccessful, what versions of gcc and openssl?

**Comment by simon987** [comment]
> The compilation has been unsuccessful

Please send the logs

> what versions of gcc and openssl?

It should work with any recent version of GCC. OpenSSL is not required

**Comment by w796933** [comment]
cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g   -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
	-o objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o \
	/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c: In function 'verify_response':
/root/ngx_http_js_challenge_module//ngx_http_js_challenge.c:285:5: error: implicit declaration of function 'SHA1' [-Werror=implicit-function-declaration]
     SHA1((unsigned char *) response.data, response.len, md);
     ^~~~
cc1: all warnings being treated as errors
make[1]: *** [objs/Makefile:1203: objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o] Error 1
make[1]: Leaving directory '/root/nginx-1.16.1'
make: *** [Makefile:8: build] Error 2

**Comment by simon987** [comment]
thank you for checking, can you try again with the last commit?

**Comment by w796933** [comment]
now i use  last build    this error . 




/root/nginx-1.16.1/src/http/ngx_http.c:2052: multiple definition of `ngx_http_set_default_types'
objs/src/http/ngx_http.o:/root/nginx-1.16.1/src/http/ngx_http.c:2052: first defined here
objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o: In function `ngx_http_merge_types':
/root/nginx-1.16.1/src/http/ngx_http.c:1988: multiple definition of `ngx_http_merge_types'
objs/src/http/ngx_http.o:/root/nginx-1.16.1/src/http/ngx_http.c:1988: first defined here
objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o: In function `ngx_http_init_locations':
/root/nginx-1.16.1/src/http/ngx_http.c:744: multiple definition of `ngx_http_module'
objs/src/http/ngx_http.o:/root/nginx-1.16.1/src/http/ngx_http.c:632: first defined here
objs/addon/ngx_http_js_challenge_module/ngx_http_js_challenge.o: In function `ngx_http_cmp_locations':
/root/nginx-1.16.1/src/http/ngx_http.c:928: multiple definition of `ngx_http_html_default_types'
objs/src/http/ngx_http.o:/root/nginx-1.16.1/src/http/ngx_http.c:693: first defined here
collect2: error: ld returned 1 exit status
make[1]: *** [objs/nginx] Error 1
make[1]: Leaving directory `/root/nginx-1.16.1'
make: *** [build] Error 2

**Comment by simon987** [comment]
Are your currently running nginx.1.16.1 (check with nginx -v) ? 
And are you using the `build.sh` script?

**Comment by w796933** [comment]
http://205.185.126.131/   i build  pass   .  so  thanks .  nice day .

**Comment by w796933** [comment]
sir  is use   nginx   Reverse proxy .  and enabled   js  challenge .  but  http  get  fllod   pass   source   server . 


server {
    js_challenge on;
    js_challenge_secret "xiang";
    js_challenge_html /usr/local/nginx/conf/js.html;
    js_challenge_bucket_duration 3600;
    js_challenge_title "Verifying your browser...";


  set $forward_scheme http;
  set $server         "5.135.155.13";
  set $port           80;
  access_log off;

  listen 80;
 

  server_name 172.16.233.212;

   location /  {
     js_challenge_bucket_duration 600;
   add_header       X-Served-By $host;
proxy_set_header Host $host;
proxy_set_header X-Forwarded-Scheme $scheme;
proxy_set_header X-Forwarded-Proto  $scheme;
proxy_set_header X-Forwarded-For    $remote_addr;
proxy_pass       $forward_scheme://$server:$port;
  }
}

**Comment by simon987** [comment]
>  but http get fllod pass source server

Could you rephrase that? I don't understand what you mean.

If you're asking why you're getting this `404 Site x.x.x.x is not served on this interface` that's because you probably need to change the `$host` header to match your configuration

**Comment by w796933** [comment]
Unable to resist HTTP Flood attacks


Malicious requests will still be delivered to the backend


 you can   use  wrk   or  ab  test it .

**Comment by simon987** [comment]
Sorry about that, can you tell me if you still have this issue again with the last version?

**Comment by w796933** [comment]
yes . i use  last  version .   you can  use  wrk  test it .

**Comment by simon987** [comment]
> can you tell me if you still have this issue again with the last version?

@w796933 I think you misunderstood, I made a new fix in the latest commit and I'm asking you to test it

**Comment by w796933** [comment]
okay so thanks . now is ok.

tcp       40      0 172.16.224.94:80        172.16.233.212:48614    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:44784    CLOSE_WAIT 
tcp       41      0 209.141.36.210:80       209.141.39.26:43938     CLOSE_WAIT 
tcp       41      0 209.141.36.210:80       209.141.39.26:43704     CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:46800    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:45784    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:51438    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:51024    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:45110    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:50134    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:50316    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:50632    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:45424    CLOSE_WAIT 
tcp        0      0 209.141.36.210:80       209.141.39.26:45632     ESTABLISHED
tcp       41      0 209.141.36.210:80       209.141.39.26:44356     CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:43104    CLOSE_WAIT 
tcp       41      0 209.141.36.210:80       209.141.39.26:42768     CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:49864    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:47040    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:44034    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:43372    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:50688    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:44874    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:45810    CLOSE_WAIT 
tcp       41      0 209.141.36.210:80       209.141.39.26:44130     CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:50744    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:50176    CLOSE_WAIT 
tcp       41      0 209.141.36.210:80       209.141.39.26:42994     CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:51556    CLOSE_WAIT 
tcp       40      0 172.16.224.94:80        172.16.233.212:51092    CLOSE_WAIT 





If the js challenge fails, nginx will return a 444 status code？？ I see a js failed link, tcp did not establish a link

**Comment by w796933** [comment]
http://209.141.36.210/  is my  build ,chrome   and  firfox    No problem to visit  .  but  ie    and   safari  js   no pass .

**Comment by w796933** [comment]
sir ?
